case.go

```diff
@@ -0,0 +1,39 @@
+package main
+
+type TestCase struct {
+	Client string
+	Params map[string]string
+}
+
+func Collect(matrix Matrix) []TestCase {
+
+	cases := make([]TestCase, 0)
+	for _, client := range matrix.Clients {
+		clientTestCases := createCasesFromClient(client, matrix, 0, make(map[string]string))
+		cases = append(cases, clientTestCases...)
+	}
+
+	return cases
+}
+
+func createCasesFromClient(client string, matrix Matrix, index int, params map[string]string) []TestCase {
+
+	if index == len(matrix.ParamKeys) {
+		copy := make(map[string]string)
+		for k, v := range params {
+			copy[k] = v
+		}
+		return []TestCase{{
+			Client: client,
+			Params: copy,
+		}}
+	}
+
+	var testCases []TestCase
+	key := matrix.ParamKeys[index]
+	for _, p := range matrix.Params[key] {
+		params[key] = p
+		testCases = append(testCases, createCasesFromClient(client, matrix, index+1, params)...)
```

Comment:
You don't have to pass the indexes around. Slicing an existing slice doesn't cause allocs. Once you have a single dimension type, you can declare:

     func createCasesForClient(client string, dims []Dimension, params map[string]string) []TestCase {
      // ...
    }

And keep most of the rest of the logic the same.